docs/nnx/filters_guide.md

```diff
@@ -0,0 +1,176 @@
+    self.a = nnx.Param(0)
+    self.b = nnx.BatchStat(True)
+
+foo = Foo()
+
+graphdef, params, batch_stats = nnx.split(foo, nnx.Param, nnx.BatchStat)
+
+print(f'{params = }')
+print(f'{batch_stats = }')
+```
+
+Here `nnx.Param` and `nnx.BatchStat` are used as Filters to split the model into two groups: one with the parameters and the other with the batch statistics. However this begs the questions:
+
+* What is a Filter?
+* Why are types, such as `Param` or `BatchStat`, Filters?
+* How is `State` grouped / filtered?
+
++++
+
+## The Filter Protocol
+
+In general Filter are predicate functions of the form:
+
+```python
+
+def f(path: tuple[Key, ...], value: Any) -> bool:
+
+```
+where `Key` is a hashable and comparable type, `path` is a tuple of `Key`s representing the path to the value in a nested structure, and `value` is the value at the path. The function returns `True` if the value should be included in the group and `False` otherwise.
+
+Types are obviously not functions. The reason why types are used as Filters is because internally they are converted to functions. For example, `Param` is roughly converted to a function like this:
+
+```{code-cell}
+def is_param(path, value) -> bool:
+  return isinstance(value, nnx.Param) or (
+    hasattr(value, 'type') and issubclass(value.type, nnx.Param)
+  )
+```
+
+Such a function matches any value that is an instance of `Param` or any value that hash a `type` attribute that is an instance of a subclass of `Param`. Internally NNX uses `OfType` which creates a callable of the form for a given type.
+
+```{code-cell}
+is_param = nnx.OfType(nnx.Param)
+
+print(f'{is_param((), nnx.Param(0)) = }')
+print(f'{is_param((), nnx.VariableState(type=nnx.Param, value=0)) = }')
+```
+
+## The Filter DSL
+
+To avoid users having to create these functions, NNX exposes a small
+DSL, formalized as the `nnx.filterlib.Filter` type, which lets users for common cases
+pass certain non-callable types such as types, booleans, ellipsis, tuples/lists, etc,
+and converts them to the appropriate callable internally.
+
+Here is a list of all the callable Filters included in NNX and their DSL literals (when available):
+
+
+| Literal | Callable | Description |
+|--------|----------------------|-------------|
+| `...` | `Everything()` | Matches all values |
+| `None` | `Nothing()` | Matches no values |
+| `True` | `Everything()` | Matches all values |
+| `False` | `Nothing()` | Matches no values |
+| `type` | `OfType(type)` | Matches values that are instances of `type` or have a `type` attribute that is an instance of `type` |
+| | `PathContains(key)` | Matches values that have an associated `path` that contains the given `key` |
+| `'{filter}'` | `WithTag('{filter}')` | Matches values that have string `tag` attribute equal to `'{filter}'` |
+| `(*filters)` | `Any(*filters)` | Matches values that match any of the inner `filters` |
+| `[*filters]` | `Any(*filters)` | Matches values that match all of the inner `filters` |
```

Comment:
Is some part of this grid duplicated? Not sure I understood it fully but the tuple is "any" while the list is "all"? This feels a bit confusing...

Reply:
Merged them into 1 row to clarify that both tuples and lists are converted to `Any`